sdk/search/search-documents/src/searchIndexingBufferedSender.ts

```diff
@@ -1,133 +1,432 @@
   /**
    * Detach Batch Failed Event
    *
    * @param event - Event to be emitted
    * @param listener - Event Listener
    */
-  off(event: "batchFailed", listener: (e: RestError) => void): void;
+  public off(event: "batchFailed", listener: (e: RestError) => void): void;
+  public off(
+    event: "batchAdded" | "beforeDocumentSent" | "batchSucceeded" | "batchFailed",
+    listener: (e: any) => void
+  ): void {
+    this.emitter.removeListener(event, listener);
+  }
+
+  private isBatchReady(): boolean {
+    return this.batchObject.actions.length >= this.initialBatchActionCount;
+  }
+
+  private async internalFlush(force: boolean, options: OperationOptions = {}): Promise<void> {
+    if (force || (this.autoFlush && this.isBatchReady())) {
+      // Split it
+      const actions: IndexDocumentsAction<T>[] = this.batchObject.actions;
+      this.batchObject = new IndexDocumentsBatch<T>();
+      while (actions.length > 0) {
+        const actionsToSend = actions.splice(0, this.initialBatchActionCount);
+        await this.submitDocuments(actionsToSend, options);
+      }
+    }
+  }
+
+  private async submitDocuments(
+    actionsToSend: IndexDocumentsAction<T>[],
+    options: OperationOptions,
+    retryAttempt: number = 1
+  ): Promise<void> {
+    try {
+      for (const action of actionsToSend) {
+        this.emitter.emit("beforeDocumentSent", action);
+      }
+      const result = await this.client.indexDocuments(
+        new IndexDocumentsBatch<T>(actionsToSend),
+        options
+      );
+      // raise success event
+      this.emitter.emit("batchSucceeded", result);
+    } catch (e) {
+      if (e.code && e.code === "413" && actionsToSend.length > 1) {
+        // Cut the payload size to half
+        const splitActionsArray = [
+          actionsToSend.slice(0, actionsToSend.length / 2),
+          actionsToSend.slice(actionsToSend.length / 2, actionsToSend.length)
+        ];
+        this.initialBatchActionCount = splitActionsArray[0].length; // So, we do not want 413 happening again and again
+        for (const actions of splitActionsArray) {
+          this.submitDocuments(actions, options);
+        }
+      } else if (this.isRetryAbleError(e) && retryAttempt <= this.maxRetriesPerAction) {
+        // Exponentially increase the delay each time
+        const exponentialDelay = this.throttlingDelayInMs * Math.pow(2, retryAttempt);
+        // Don't let the delay exceed the maximum
+        const clampedExponentialDelay = Math.min(this.maxThrottlingDelayInMs, exponentialDelay);
+        // Allow the final value to have some "jitter" (within 50% of the delay size) so
+        // that retries across multiple clients don't occur simultaneously.
+        const delayWithJitter =
+          clampedExponentialDelay / 2 + getRandomIntegerInclusive(0, clampedExponentialDelay / 2);
+        await delay(delayWithJitter);
+        this.submitDocuments(actionsToSend, options, retryAttempt + 1);
```

Comment:
I think we're missing an `await` here, right? Otherwise retry attempt success/failure gets orphaned.

Reply:
Why do we have an await here? It is similar to the retry logic in line 421. We are not awaiting there. So, why will we await here? In case of success or failures, the SDK user will be notified through events (and error will be thrown in case of error). Do you still think we need an await here?